refactor(inputs): use pseudo-private custom properties

dancormier opened the pull request:
This PR refactors component styling to use pseudo-private custom properties for the following components (completed refactors checked):

- [x] `.s-check-control`
- [x] `.s-check-group`
- [x] `.s-checkbox`
- [x] `.s-input`
- [x] `.s-input-fill`
- [x] `.s-input-icon`
- [x] `.s-input-message`
- [x] `.s-select`
- [x] `.s-radio`
- [x] `.s-textarea`
- [x] Validation state classes
    - [x] `.is-disabled`
    - [x] `.is-readonly`
    - [x] `.has-success`
    - [x] `.has-error`
    - [x] `.has-warning`

> **Note**
> This PR should result in no visual changes to the above components.

All of these components are being refactored at once because their styling is currently tightly-coupled. This PR will decouple styling where ever it's logical.

### Deprecations and removals

This PR gets rid of the one and only reference to `.s-input.has-focus` in Stacks. I only found it used once in Core and that was in our developer theme testing page.

dancormier committed:
refactor(inputs): use pseudo-private custom properties

dancormier committed:
Move s-description to standalone file

dancormier committed:
Move s-checkbox, s-radio, and similar to standalone file

dancormier committed:
Move @input-focus-style to helpers

dancormier committed:
Cleanup input.less

dancormier committed:
Organize global modifier classes

dancormier committed:
More organization of global modifiers

dancormier committed:
Remove dupe global modifier rules

dancormier committed:
Move disabled fieldset button styling to .s-btn

dancormier committed:
Move disabled fieldset link, anchor styling to links.less

dancormier committed:
Move fieldset[disabled] styles

dancormier committed:
Replace .appearance() mixin with appearance property

We're WELL PAST the point of wide-spread, non-vendor-prefixed `appearance` support. Keeping the prefixed rules around when non-prefixed is not also referenced in the off chance that we're intentionally only targeting webkit and/or mozilla.
https://caniuse.com/?search=appearance

dancormier committed:
Light checkbox and radio style organization

dancormier committed:
Whoops, those styles are the base fieldset reset

dancormier committed:
Base .s-checkbox, .s-radio refactor

dancormier committed:
A little more checkbox, radio cleanup

dancormier committed:
Unify checkbox, radio PPCPs; refactor global modifier styles

dancormier committed:
Fix readonly focus, high contrast styling

b-kelly commented on lib/css/base/reset-normalize.less:
```diff
@@ -311,7 +311,9 @@ button:-moz-focusring,
  */
 
 fieldset {
-  padding: 0.35em 0.75em 0.625em;
```
I'd like to be cautious tweaking either of the resets, as consumers may be depending on those for _other_ styles.

I'd recommend making this change in your component css instead if it is required.

dancormier replied:
Ah, good catch! I [restored this style within the the reset and created fieldset.less](https://github.com/StackExchange/Stacks/pull/1184/commits/c364106cbbc6c591917756f0da7ed5bfc71796bb) which includes the styles previously in `inputs.less`. I chose the standalone `lib/css/base/fieldset.less` file since the style is not specific to any specific component.

b-kelly commented on lib/css/components/description.less:
```diff
@@ -0,0 +1,9 @@
+.s-description {
```
Good call pulling this one out. We use this "component" in a few places, but never really document it.

b-kelly commented on lib/css/exports/global-modifiers.less:
```diff
@@ -0,0 +1,7 @@
+.is-disabled,
```
Imo, we should inline this one style anywhere it is used. This is something that could bite us in the rear later on, especially if we're including it in `exports` (essentially making it part of our public LESS api)

dancormier replied:
Ah, I didn't really consider the implications of including this in our `/exports/` directory. Yeah, this really isn't the best home for these classes.

The issue is that these classes have this rule applied without any parent/sibling classes applied, so I'm not sure where the best permanent home would be (besides a file like this one). My first instinct is to move this file to `lib/css/atomic` since they're applied _kinda_ like atomic classes, but really they're modifiers 🤷‍♂️ Any thoughts?

ondrejkonec replied:
Hi guys! Quick note: as you know, this State rules comes from SMACSS approach (http://smacss.com/book/type-state) and they are mostly tied to changing behavior using JS. I think they are good for naming component states (you can see very quickly what is going on in CSS and what variant it is) and IMHO they don't need to be in a separate file because they have no general meaning.

ondrejkonec replied:
If you plan to use for e.g. "is-hidden" or "is-invisible" that can be used everywhere, I think some /atomic class folder should work here.

dancormier replied:
@giamir @b-kelly any ideas wrt [my comment](https://github.com/StackExchange/Stacks/pull/1184#discussion_r1038536330)?

giamir replied:
I haven't reviewed in details this PR but at a first glance I also find odd to see these classes globally available across the whole document.  As a rule of thumbs we should always think twice before exposing more API surface to consumers because we are implicitly making the system harder for us to change/evolve.
I would prefer if we could keep those classes scoped to the component that needs them (similar to what Ben suggested). This will create a bit of duplication but I would not mind to prioritize encapsulation over performance in this context.
If we have instances where these classes are just used in random contexts then we could put the rules in a place where it's clear we will deprecate them in the next major release (maybe a `pending-deprecation.less` file?) How much work do you foresee in Core for replacing these classes usage?
By the way the deprecation strategy ADR and the css dev bundle experiment is still in my todo list. I should probably create an issue for that but won't be able to tackle it before next year.
I hope this rambling is somewhat useful. 🙂

giamir replied:
>  If we have instances where these classes are just used in random contexts then we could put the rules in a place where it's clear we will deprecate them in the next major release (maybe a pending-deprecation.less file?) How much work do you foresee in Core for replacing these classes usage?

@dancormier @b-kelly Do you think is possible to deprecate those classes in our next major release? How painful would that be?

dancormier replied:
> How much work do you foresee in Core for replacing these classes usage?

I think it'd be a pretty heavy lift to replace those classes in Core. There's lots of usage of `.is-disabled` and `.has-error` that would need to be replaced. Often, these classes are set with JS conditionally, so time would be needed to understand context and prevent regressions.

> How painful would that be?

I think it would be pretty painful and honestly I feel like this is something we should punt on. In the context of this PR, nothing would materially change for consumers wrt these `.is-` and `.has-` classes. I'm good with being overruled if I'm being too pessimistic, but felt I should voice my opinion.

dancormier replied:
Outside of the deprecation consideration, where could be the best place for these styles @giamir @b-kelly? `lib/css/atomic`?

b-kelly replied:
Looking at some of the other code, we may need to keep around an e.g. `input-utils.less` file for mixins. These could likely go in there as well.

b-kelly reviewed (commented):
I haven't reviewed visually yet, but I love that you're splitting these components off into their own files.

I have a few small nits that I think could be easily rectified before merging.

dancormier committed:
Merge branch 'develop' into dcormier/input-refactor

dancormier committed:
Restore fieldset reset rules; add fieldset.less

dancormier committed:
Cleanup stacks-static.less

dancormier committed:
Merge branch 'develop' into dcormier/input-refactor

dancormier committed:
Cleanup

dancormier committed:
Merge branch 'develop' into dcormier/input-refactor

dancormier committed:
Add Netlify redirect from docs to readme

https://github.com/StackExchange/Stacks/pull/1211#issuecomment-1349519151

dancormier committed:
Revert "Add Netlify redirect from docs to readme"

This reverts commit 51bbbe492a11d4375cd08482f4c5a0d3f692dacb.

dancormier committed:
Merge branch 'develop' into dcormier/input-refactor

dancormier committed:
Move style from links.less to link.less

dancormier committed:
Merge branch 'develop' into dcormier/input-refactor

dancormier committed:
Add flex-shrink: 0 to checkboxes and radios

dancormier committed:
Merge branch 'develop' into dcormier/input-refactor

dancormier committed:
make .s-input-icon styling independent of modifier classes

dancormier committed:
turns out unset doesn't work for border-radius here

dancormier committed:
make linter happy

dancormier committed:
Merge branch 'develop' of https://github.com/StackExchange/Stacks into dcormier/input-refactor

b-kelly commented on lib/css/components/link.less:
```diff
@@ -68,6 +68,13 @@ a,
     }
 
+    // MODIFIERS
+    fieldset[disabled] & {
```
Move fieldset stuff all into `fieldset.less`. Two ways to think about it:

1. Styles are targeting the child, put them in the child file
2. Styles are in the context of the parent, put them in the parent file

This is not a hill I'm willing to die on, just food for thought.

b-kelly replied:
This was originally applied to `a,.s-btn,.s-link`. We're covering `a, .s-link`, but not `.s-btn` I believe.

https://github.com/StackExchange/Stacks/blob/develop/lib/css/components/inputs.less#L91-L93

dancormier replied:
> Styles are in the context of the parent, put them in the parent file

I opted for option 2 in the spirit of component styles being self-contained. Basically, my goal is that anyone trying to see what could be affecting a given component's styling can look in a single file. Big picture, if we every decided to get rid of or rename a component, all the styles would (optimally) live in the component's `.less` file and hopefully be slightly easier to handle without leaving crust around our codebase.

---

> This was originally applied to a,.s-btn,.s-link. We're covering a, .s-link, but not .s-btn I believe.

I'm not sure how we missed it, but I added this style to [buttons.less](https://github.com/StackExchange/Stacks/blob/43e3ebe364639827fc12b7b0a525c388c56f1d77/lib/css/components/buttons.less#L60) in https://github.com/StackExchange/Stacks/commit/939f1025b0e3a4b108637f3cb9e1adf3c6b2f5ca

b-kelly commented on lib/css/stacks-static.less:
One note to make here is that when we share Stacks from SO to other projects/sites (e.g. StackExchange.com) _only_ the static styles are included (for legacy theming reasons that don't matter any longer now that we're using css variables after we remove `stacks-dynamic.less`).

No action necessary.

b-kelly commented on lib/css/components/checkboxes-radios.less:
```diff
@@ -0,0 +1,181 @@
+    }
+
+    // STATE
+    &::-ms-check { // TODO do we still need this? I don't think so, but help me Windows peeps!
```
Double checked this - this was to support Edge Legacy and IE (which we do not support). This is safe to remove.

b-kelly commented on lib/css/components/checkboxes-radios.less:
```diff
@@ -0,0 +1,181 @@
+    background-color: var(--_ch-bg);
+    border: var(--_ch-baw) solid var(--_ch-bc);
+
+    appearance: none;
+    cursor: pointer;
+    font-size: inherit;
+    height: 1em;
+    margin: 0; // A guard against Core's default margins
+    outline: 0;
+    vertical-align: middle;
+    width: 1em;
```
Looks like these styles were formally gated behind a `supports` media query? Still necessary with our supported browser list?

https://github.com/StackExchange/Stacks/blob/develop/lib/css/components/inputs.less#L289

b-kelly replied:
Double checking the support list, we're widely in the clear here (Chrome 1, Safari 3, Firefox 54)

b-kelly commented on lib/css/components/checkboxes-radios.less:
```diff
@@ -0,0 +1,181 @@
+        }
+    });
+
+    .dark-mode({
```
PR NOTE: Review visually, added style

dancormier replied:
This resolves a bug where the center dot in checked radio elements would fill with the wrong color in dark mode.

### Before

![image](https://user-images.githubusercontent.com/647177/216161223-267f4b0d-8ac9-42eb-9094-19f08e294876.png)

### After
![image](https://user-images.githubusercontent.com/647177/216161184-2ab92f6c-7ec0-40b1-8971-a4e18d070ee5.png)

b-kelly commented on lib/css/components/checkboxes-radios.less:
```diff
@@ -0,0 +1,181 @@
+    &:checked {
+        --_ch-baw: 0.30769231em;
+        --_ch-bc: var(--theme-secondary-400);
+        --_ch-bg: var(--white); // TODO seems to not change anything. Investigate
```
Can we clean up this TODO?

b-kelly commented on lib/css/components/checkboxes-radios.less:
```diff
@@ -0,0 +1,181 @@
+
+.s-checkbox,
+.s-radio:not(:checked) {
+    .has-error &,
```
Note to self - this could likely be a mixin that is shared amongst all inputs. Doing so would help to deduplicate values

dancormier replied:
Added `.validation-states` mixin here: https://github.com/StackExchange/Stacks/pull/1184/commits/2086b63f5460027b9d19bebb6c2283bf2060ff18

b-kelly commented on lib/css/components/checkboxes-radios.less:
```diff
@@ -0,0 +1,181 @@
+    }
+
+    // CHILD ELEMENTS
+    legend.s-label { // TODO HACK? <legend> isn't respecting gap...
```
optional: Personal preference to keep the width low (and sometimes putting on the same line as an opening bracket screws with formatting).

Ignore if you like.

```suggestion
    // TODO HACK? <legend> isn't respecting gap...
    legend.s-label {
```

b-kelly commented on lib/css/components/select.less:
```diff
@@ -0,0 +1,194 @@
+.s-select {
+    --_se-arrow-bc: currentColor transparent;
+    --_se-arrow-size: var(--su-static4); // Constant // TODO create name/styling convension for constants
```
Cleanup TODO?

b-kelly commented on lib/css/components/select.less:
```diff
@@ -0,0 +1,194 @@
+        --_se-arrow-bc: var(--bc-darker) transparent;
+    }
+
+    &&__sm {
```
Sizing could also likely benefit greatly from a mixin that is shared between all inputs

dancormier replied:
I added a mixin in this commit https://github.com/StackExchange/Stacks/pull/1184/commits/718c52ee7e194d55b04f313349f61ad91613d047. I went pretty programmatic since we're using pseudo-private custom properties. It came out looking a little wacky, though I think the mixin is pretty effective. Let me know if it's too wacky (or not wacky enough).

b-kelly commented on lib/css/components/select.less:
```diff
@@ -0,0 +1,194 @@
+    }
+
+    // CHILD ELEMENTS
+    &,
```
Added - double check visually. To make this more safe we should probably do:

```suggestion
    select&,
```

b-kelly commented on lib/css/components/select.less:
```diff
@@ -0,0 +1,194 @@
+    // CHILD ELEMENTS
+    &,
+    & > select {
+        &::-webkit-contacts-auto-fill-button {
```
These may need to be pulled out into a mixin since it's fairly complicated and is shared amongst a few components

b-kelly commented on lib/css/components/select.less:
```diff
@@ -0,0 +1,194 @@
+            text-shadow: 0 0 0 #000;
+        }
+
+        &::-ms-expand {
```
Double check - can almost assuredly remove (Edge Legacy)

dancormier replied:
I'm removing it. It's hard to _even find_ documentation on it, so I'm pretty we can trash it. Best docs I found https://udn.realityripple.com/docs/Archive/Web/CSS/::-ms-expand

b-kelly commented on lib/css/components/select.less:
```diff
@@ -0,0 +1,194 @@
+        }
+
+        // CHILD ELEMENTS
+        &::-moz-focus-inner {
+            outline: none !important;
+        }
+
+        &:-moz-focusring {
```
PR Note - double check, necessary or is this on a standards track now?

dancormier replied:
AFAICT, we're good to dump this. I tested and saw no difference in FF with or without this rule.

![image](https://user-images.githubusercontent.com/647177/215874873-9f470dea-ef60-44f8-b13a-377befbe644b.png)

Furthermore, [MDN says](https://developer.mozilla.org/en-US/docs/Web/CSS/:-moz-focusring):
> Non-standard: This feature is non-standard and is not on a standards track. Do not use it on production sites facing the Web: it will not work for every user. There may also be large incompatibilities between implementations and the behavior may change in the future.

b-kelly commented on lib/css/components/select.less:
```diff
@@ -0,0 +1,194 @@
+        }
+
+        // INTERACTION
+        &:focus {
```
Potentially abstract back out into a mixin?

dancormier replied:
This would only partially dedupe in `.s-select` and `.s-input`/`.s-textarea`, so I don't think an abstraction would be worth it here.

b-kelly commented on lib/css/components/toggle-switches.less:
```diff
@@ -5,6 +5,14 @@
     --_ts-multiple-bg: unset;
     --_ts-multiple-fc: var(--black-500);
 
+    fieldset[disabled] & {
+        &,
+        & label {
```
Optional: simplification? removes a level of indentation, but don't know if it is cleaner

```
    fieldset[disabled] &, fieldset[disabled] & label {
```

dancormier replied:
I honestly prefer what we have already, but I think we'd be wise to look out for ugly nested `&`. I've been bit by misinterpretation there and I'll keep in mind where this style might be a little more confusing than it's worth.

b-kelly reviewed (changes requested):
Code only review, will follow up with a visual review

dancormier committed:
Update lib/css/components/checkboxes-radios.less

Co-authored-by: Ben Kelly <b-kelly@users.noreply.github.com>

dancormier committed:
Merge branch 'develop' into dcormier/input-refactor

dancormier committed:
Remove legacy -ms-check css rules

Resolves https://github.com/StackExchange/Stacks/pull/1184/files#r1089385821

dancormier committed:
Cleaned up resolved TODO

We set the checkbox and radio element backgrounds to --white so they don't blend into non-white backgrounds. Addresses https://github.com/StackExchange/Stacks/pull/1184/files#r1089394539

dancormier committed:
Remove unneeded TODO

Resolves https://github.com/StackExchange/Stacks/pull/1184/files#r1089402152

dancormier committed:
Prepend select selector to prevent dupe rule application

Co-authored-by: Ben Kelly <b-kelly@users.noreply.github.com>

dancormier committed:
Remove :-moz-focusring select selector

Resolves https://github.com/StackExchange/Stacks/pull/1184/files#r1089410265

dancormier committed:
Remove -ms-expand select selector

Resolves https://github.com/StackExchange/Stacks/pull/1184/files#r1089409854

dancormier committed:
Merge branch 'dcormier/input-refactor' of https://github.com/StackExchange/Stacks into dcormier/input-refactor

dancormier committed:
refactor: add mixin for size styles

Addresses https://github.com/StackExchange/Stacks/pull/1184#discussion_r1089403265

dancormier committed:
refactor: add webkit-autofill mixin

Addresses https://github.com/StackExchange/Stacks/pull/1184#discussion_r1089408060

dancormier committed:
refactor: add input-utils.less; input-states + validation-states mixins

Addresses

dancormier committed:
Merge branch 'develop' into dcormier/input-refactor

dancormier committed:
Merge branch 'develop' into dcormier/input-refactor

dancormier committed:
Delete global-modifiers.less

dancormier committed:
Fix .size-styles less function
